File: 2415. Reverse Odd Levels of Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def reverseOddLevels(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        queue = deque([root])
        temp = 0
        parity = True
        while queue:
            n = len(queue)
            temp = []
            for i in range(n):
                a = queue.popleft()
                if a.left:
                    temp.append(a.left)
                    queue.append(a.left)
                if a.right:
                    temp.append(a.right)
                    queue.append(a.right)
            if parity:  
                j = len(temp)-1  
                for i in range(len(temp)//2):
                    if temp[j] and temp[i]:
                        temp[j].val,temp[i].val = temp[i].val,temp[j].val
                    j -= 1
            parity = not parity    
        return root


        

File: 2415. Reverse Odd Levels of Binary Tree/solution.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;
use core::mem::swap;
impl Solution {
    pub fn reverse_odd_levels(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut queue:VecDeque<Option<Rc<RefCell<TreeNode>>>> = VecDeque::new();
        queue.push_back(root.clone());
        let mut temp:Vec<Option<Rc<RefCell<TreeNode>>>> = vec![];
        let mut parity = true;
        while !queue.is_empty(){
            let range = queue.len();
            let mut temp = vec![];
            for _ in 0..range{
                   if let Some(n) = queue.pop_front(){

                    let mut node = n.as_ref().unwrap().borrow();
                    if let Some(left) = node.left.as_ref(){
                        temp.push(left.clone());
                        queue.push_back(Some(left.clone()));
                    } 
                    if let Some(right) = node.right.as_ref(){
                        temp.push(right.clone());
                        queue.push_back(Some(right.clone()));
                    } }}
                
             if parity == true{
                let mut right = temp.len()-1;
                for left in 0..temp.len()/2{
                    swap(&mut temp[left ].borrow_mut().val,
                         &mut temp[right].borrow_mut().val);
                    right -= 1;     
                }
               
            }  parity = !parity;
            
        } root



    }}
